### PR TITLE
vector: add `package` Option

### PR DESCRIPTION
**Pull request by kai-tub**
###### Description of changes

Apply `package` option to `vector` package and to the _test_ as well.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/logging/vector.nix [hunk 1]
```diff
@@ -8,6 +8,13 @@ in
   options.services.vector = {
     enable = mkEnableOption (lib.mdDoc "Vector");
 
+    package = mkOption {
+      type = types.package;
+      default = pkgs.vector;
+      defaultText = literalExpression "pkgs.vector";
+      description = lib.mdDoc "The package to use for vector";
+    };
+
```
Comment (IndeedNotJames) [inline nixos/modules/services/logging/vector.nix:17]:
```suggestion
    package = mkPackageOptionMD pkgs "vector" { };
```

File: nixos/modules/services/logging/vector.nix [hunk 2]
```diff
@@ -47,7 +54,7 @@ in
             '';
         in
         {
-          ExecStart = "${pkgs.vector}/bin/vector --config ${validateConfig conf}";
+          ExecStart = "${cfg.package}/bin/vector --config ${validateConfig conf}";
```
Comment (IndeedNotJames) [inline nixos/modules/services/logging/vector.nix:57]:
```suggestion
          ExecStart = "${getExe cfg.package} --config ${validateConfig conf}";
```

File: nixos/tests/vector.nix [hunk 3]
```diff
@@ -13,6 +13,7 @@ with pkgs.lib;
       services.vector = {
         enable = true;
         journaldAccess = true;
+        package = pkgs.vector;
```
Comment (IndeedNotJames) [inline nixos/tests/vector.nix:16]:
```suggestion
```

Not needed.

**Review by emilylange — CHANGES_REQUESTED** [review]

**Comment by kai-tub** [comment]
Ok @IndeedNotJames, I think I've integrated your comments.
Sorry, first-time contributor, and I wasn't aware of these options. :)

I've followed https://nixos.org/manual/nixpkgs/stable/#submitting-changes-hotfixing-pull-requests and, according to it rebased on master and squashed my commits to the original commit.

**Review by jian-lin — APPROVED** [review]

**Comment by IndeedNotJames** [comment]
Hi, sorry, @kai-tub^^
I was a bit in a hurry and overlooked the *First-time contributor* badge.

Welcome to nixpkgs nonetheless! :sweat_smile: 
I can elaborate my reasoning behind each review suggestion, if you want. Just let me know :)

The one ofborg ci failure (https://logs.ofborg.org/?key=nixos/nixpkgs.231096&attempt_id=9f3810eb-cb86-4d71-82cd-bd2332d75984) should be unrelated to your change. That `aarch64-darwin` runner probably had a little hiccup, that's all:
```
error: cannot connect to socket at '/nix/var/nix/gc-socket/socket': No such file or directory
```
// rerun
@ofborg test vector

Not requesting a change, but just so you know in the future:
Next time you work on a nixos module (not nixpkgs package), refer to it using the `nixos/<name>` prefix instead[^1][^2] :)
E.g. `nixos/vector: add package option`

> I've followed https://nixos.org/manual/nixpkgs/stable/#submitting-changes-hotfixing-pull-requests and, according to it rebased on master and squashed my commits to the original commit.

Awesome! Thanks :)
The guidelines and manual(s) can be a bit confusing ¯\\\_(ツ)_/¯

[^1]: https://nixos.org/manual/nixpkgs/stable/#submitting-changes-making-patches
[^2]: https://github.com/NixOS/nixpkgs/blob/ffd4978ea9236687d0bf74c3758810eb7d592572/CONTRIBUTING.md#submitting-changes

**Review by emilylange — APPROVED** [review]
The fail in https://logs.ofborg.org/?key=nixos/nixpkgs.231096&attempt_id=5862bf17-ca09-48ef-976e-687fac21567d is, again, unrelated.
Just a fairly old race condition. (Odds are against us today :sweat_smile:)

```patch
diff --git a/nixos/tests/vector.nix b/nixos/tests/vector.nix
index 9309f6a14dd..a55eb4e012c 100644
--- a/nixos/tests/vector.nix
+++ b/nixos/tests/vector.nix
@@ -31,7 +31,7 @@ with pkgs.lib;
     # ensure vector is forwarding the messages appropriately
     testScript = ''
       machine.wait_for_unit("vector.service")
-      machine.succeed("test -f /var/lib/vector/logs.log")
+      machine.wait_for_file("/var/lib/vector/logs.log")
     '';
   };
 }
```

Fixes it.

Up to @kai-tub if they want to add another commit titled `nixosTests.vector: fix race condition` :)

Not blocking.

**Comment by IndeedNotJames** [comment]
@ofborg test vector

**Comment by kai-tub** [comment]
> Hi, sorry, @kai-tub^^ I was a bit in a hurry and overlooked the _First-time contributor_ badge.
> 
> Welcome to nixpkgs nonetheless! :sweat_smile: I can elaborate my reasoning behind each review suggestion, if you want. Just let me know :)

Hey @IndeedNotJames, no worries.
I bet there is quite a lot going on in this repository and it is probably easy to miss the badge (didn't even know there was such a badge)

I think I understood your comments, once I found the documentation. Or rather found the comments in `nixpkgs` directly, as I couldn't seem to find the `getExe` documentation.

> Just a fairly old race condition. (Odds are against us today :sweat_smile:)

Thank you very much for investigating the issue!
The flakey tests were actually the main reason why I wanted this option, as `vector` was frequently failing in CI...

> E.g. nixos/vector: add package option

Ah, yes. I missed that part. I will do it better next time. :)
But one remaining question:

> The guidelines and manual(s) can be a bit confusing ¯\_(ツ)_/¯

Does this mean that I shouldn't have force pushed?
If yes, than I can also fix the commit message if you don't mind running CI yet again :sweat_smile:

**Comment by IndeedNotJames** [comment]
> I think I understood your comments, once I found the documentation. Or rather found the comments in `nixpkgs` directly, as I couldn't seem to find the `getExe` documentation.

Yeah, I am not aware of any official documentation of nixpkgs' `lib.*` functions besides the comments above each function.

E.g. https://github.com/NixOS/nixpkgs/blob/897876e4c484f1e8f92009fd11b7d988a121a4e7/lib/meta.nix#L135-L147

But there are some community projects, like https://noogle.dev/ and https://teu5us.github.io/nix-lib.html#lib.meta.getexe, that parse those comments for the time being :)
Probably won't change any time soon.

> The flakey tests were actually the main reason why I wanted this option, as `vector` was frequently failing in CI...

Nixpkgs ofborg/hydra CI or your own?


> Ah, yes. I missed that part. I will do it better next time. :)

No worries :)
(At least from me.)

> Does this mean that I shouldn't have force pushed?

No no, all good.
I just wanted to acknowledge that you did the right thing despite the rather complex and confusing/unusual guidelines :)



> If yes, than I can also fix the commit message if you don't mind running CI yet again :sweat_smile:

IMHO, you don't _need_ to fix your commit message. But if you want, sure, go ahead :)
You can also trigger ofborg yourself.
See https://github.com/NixOS/ofborg#commands
Or just copy the `@ofborg test vector` I commented above

**Comment by kai-tub** [comment]
> But there are some community projects, like https://noogle.dev/ and https://teu5us.github.io/nix-lib.html#lib.meta.getexe, that parse those comments for the time being :)
Probably won't change any time soon.

:exploding_head: Ok, this is pretty helpful. Thanks!

> Nixpkgs ofborg/hydra CI or your own?

hydra CI and my own test once before committing the first time, as I got confused about why my change would break the package and then took a look at: https://hydra.nixos.org/job/nixos/trunk-combined/nixos.tests.vector.test1.x86_64-linux which showed me that it had issues on master. But it seems to be the one you just "fixed" :) 

> You can also trigger ofborg yourself.

Ok, let's see:

@ofborg test vector


Thanks for all the help!

**Pull request merged**